File: frontend/src/features/automations/schemas.ts
import { z } from "zod";

export const AutomationScheduleTypeSchema = z.enum(["daily"]);
export const AutomationScheduleDaySchema = z.enum(["mon", "tue", "wed", "thu", "fri", "sat", "sun"]);
export const AutomationRunStatusSchema = z.enum(["running", "success", "failed", "partial"]);
export const AutomationRunTriggerSchema = z.enum(["scheduled", "manual"]);
export const AutomationReasoningEffortSchema = z.enum(["minimal", "low", "medium", "high", "xhigh"]);

export const AutomationScheduleDaysSchema = z
  .array(AutomationScheduleDaySchema)
  .min(1)
  .max(7)
  .refine((value) => new Set(value).size === value.length, "Duplicate schedule days are not allowed");

export const AutomationScheduleSchema = z.object({
  type: AutomationScheduleTypeSchema,
  time: z.string().regex(/^\d{2}:\d{2}$/),
  timezone: z.string().min(1),
  thresholdMinutes: z.number().int().min(0).max(240).default(0),
  days: AutomationScheduleDaysSchema,
});

export const AutomationRunSchema = z.object({
  id: z.string().min(1),
  jobId: z.string().min(1),
  jobName: z.string().nullable().optional(),
  model: z.string().nullable().optional(),
  reasoningEffort: AutomationReasoningEffortSchema.nullable().optional(),
  trigger: AutomationRunTriggerSchema,
  status: AutomationRunStatusSchema,
  scheduledFor: z.string().datetime({ offset: true }),
  startedAt: z.string().datetime({ offset: true }),
  finishedAt: z.string().datetime({ offset: true }).nullable(),
  accountId: z.string().nullable(),
  errorCode: z.string().nullable(),
  errorMessage: z.string().nullable(),
  attemptCount: z.number().int().nonnegative(),
  effectiveStatus: AutomationRunStatusSchema.nullable().optional(),
  totalAccounts: z.number().int().nonnegative().nullable().optional(),
  completedAccounts: z.number().int().nonnegative().nullable().optional(),
  pendingAccounts: z.number().int().nonnegative().nullable().optional(),
  cycleKey: z.string().nullable().optional(),
});

export const AutomationJobSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  enabled: z.boolean(),
  includePausedAccounts: z.boolean().default(false),
  schedule: AutomationScheduleSchema,
  model: z.string().min(1),
  reasoningEffort: AutomationReasoningEffortSchema.nullable().optional(),
  prompt: z.string().min(1),
  accountScopeAll: z.boolean().default(true),
  accountIds: z.array(z.string().min(1)).default([]),
  nextRunAt: z.string().datetime({ offset: true }).nullable(),
  lastRun: AutomationRunSchema.nullable(),
});

export const AutomationsListResponseSchema = z.object({
  items: z.array(AutomationJobSchema).default([]),
  total: z.number().int().nonnegative().default(0),
  hasMore: z.boolean().default(false),
});

export const AutomationRunsListResponseSchema = z.object({
  items: z.array(AutomationRunSchema).default([]),
  total: z.number().int().nonnegative().default(0),
  hasMore: z.boolean().default(false),
});

export const AutomationJobFilterOptionsSchema = z.object({
  accountIds: z.array(z.string().min(1)).default([]),
  models: z.array(z.string().min(1)).default([]),
  statuses: z.array(z.string().min(1)).default([]),
  scheduleTypes: z.array(z.string().min(1)).default([]),
});

export const AutomationRunFilterOptionsSchema = z.object({
  accountIds: z.array(z.string().min(1)).default([]),
  models: z.array(z.string().min(1)).default([]),
  statuses: z.array(z.string().min(1)).default([]),
  triggers: z.array(z.string().min(1)).default([]),
});

export const AutomationRunAccountStateSchema = z.object({
  accountId: z.string().min(1),
  status: z.enum(["pending", "running", "success", "failed", "partial"]),
  runId: z.string().nullable().optional(),
  scheduledFor: z.string().datetime({ offset: true }).nullable().optional(),
  startedAt: z.string().datetime({ offset: true }).nullable().optional(),
  finishedAt: z.string().datetime({ offset: true }).nullable().optional(),
  errorCode: z.string().nullable().optional(),
  errorMessage: z.string().nullable().optional(),
});

export const AutomationRunDetailsSchema = z.object({
  run: AutomationRunSchema,
  accounts: z.array(AutomationRunAccountStateSchema).default([]),
  totalAccounts: z.number().int().nonnegative(),
  completedAccounts: z.number().int().nonnegative(),
  pendingAccounts: z.number().int().nonnegative(),
});

export const AutomationCreateRequestSchema = z.object({
  name: z.string().min(1),
  enabled: z.boolean().default(true),
  includePausedAccounts: z.boolean().optional(),
  schedule: AutomationScheduleSchema,
  model: z.string().min(1),
  reasoningEffort: AutomationReasoningEffortSchema.nullable().optional(),
  prompt: z.string().optional(),
  accountIds: z.array(z.string().min(1)),
});

export const AutomationUpdateRequestSchema = z.object({
  name: z.string().min(1).optional(),
  enabled: z.boolean().optional(),
  includePausedAccounts: z.boolean().optional(),
  schedule: AutomationScheduleSchema.optional(),
  model: z.string().min(1).optional(),
  reasoningEffort: AutomationReasoningEffortSchema.nullable().optional(),
  prompt: z.string().min(1).optional(),
  accountIds: z.array(z.string().min(1)).optional(),
});

export const AutomationDeleteResponseSchema = z.object({
  status: z.literal("deleted"),
});

export type AutomationScheduleType = z.infer<typeof AutomationScheduleTypeSchema>;
export type AutomationScheduleDay = z.infer<typeof AutomationScheduleDaySchema>;
export type AutomationRunStatus = z.infer<typeof AutomationRunStatusSchema>;
export type AutomationRunTrigger = z.infer<typeof AutomationRunTriggerSchema>;
export type AutomationReasoningEffort = z.infer<typeof AutomationReasoningEffortSchema>;
export type AutomationSchedule = z.infer<typeof AutomationScheduleSchema>;
export type AutomationRun = z.infer<typeof AutomationRunSchema>;
export type AutomationJob = z.infer<typeof AutomationJobSchema>;
export type AutomationsListResponse = z.infer<typeof AutomationsListResponseSchema>;
export type AutomationRunsListResponse = z.infer<typeof AutomationRunsListResponseSchema>;
export type AutomationCreateRequest = z.infer<typeof AutomationCreateRequestSchema>;
export type AutomationUpdateRequest = z.infer<typeof AutomationUpdateRequestSchema>;
export type AutomationDeleteResponse = z.infer<typeof AutomationDeleteResponseSchema>;
export type AutomationJobFilterOptions = z.infer<typeof AutomationJobFilterOptionsSchema>;
export type AutomationRunFilterOptions = z.infer<typeof AutomationRunFilterOptionsSchema>;
export type AutomationRunAccountState = z.infer<typeof AutomationRunAccountStateSchema>;
export type AutomationRunDetails = z.infer<typeof AutomationRunDetailsSchema>;

File: app/modules/automations/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field, field_validator

from app.modules.shared.schemas import DashboardModel

AUTOMATION_SCHEDULE_TYPES = ("daily",)
AUTOMATION_WEEKDAY_CODES = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")
AUTOMATION_RUN_STATUSES = ("running", "success", "failed", "partial")
AUTOMATION_RUN_TRIGGERS = ("scheduled", "manual")
AUTOMATION_REASONING_EFFORTS = ("minimal", "low", "medium", "high", "xhigh", "max", "ultra")

AutomationWeekday = Literal["mon", "tue", "wed", "thu", "fri", "sat", "sun"]


def _default_automation_weekdays() -> list[str]:
    return ["mon", "tue", "wed", "thu", "fri", "sat", "sun"]


class AutomationScheduleRequest(DashboardModel):
    type: str = "daily"
    time: str
    timezone: str
    threshold_minutes: int = 0
    days: list[str] = Field(default_factory=_default_automation_weekdays)


class AutomationScheduleResponse(DashboardModel):
    type: Literal["daily"]
    time: str
    timezone: str
    threshold_minutes: int
    days: list[AutomationWeekday]


class AutomationJobCreateRequest(DashboardModel):
    name: str = Field(min_length=1, max_length=200)
    enabled: bool = True
    include_paused_accounts: bool = Field(default=False, alias="includePausedAccounts")
    schedule: AutomationScheduleRequest
    model: str = Field(min_length=1)
    reasoning_effort: str | None = Field(default=None, pattern=r"(?i)^(minimal|low|medium|high|xhigh|max|ultra)$")
    prompt: str | None = Field(default=None, max_length=1000)
    account_ids: list[str] = Field(default_factory=list, max_length=128, alias="accountIds")

    @field_validator("account_ids")
    @classmethod
    def _validate_unique_account_ids(cls, value: list[str]) -> list[str]:
        normalized = [account_id.strip() for account_id in value if account_id.strip()]
        if len(set(normalized)) != len(normalized):
            raise ValueError("Duplicate account IDs are not allowed")
        return normalized


class AutomationJobUpdateRequest(DashboardModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    enabled: bool | None = None
    include_paused_accounts: bool | None = Field(default=None, alias="includePausedAccounts")
    schedule: AutomationScheduleRequest | None = None
    model: str | None = Field(default=None, min_length=1)
    reasoning_effort: str | None = Field(default=None, pattern=r"(?i)^(minimal|low|medium|high|xhigh|max|ultra)$")
    prompt: str | None = Field(default=None, max_length=1000)
    account_ids: list[str] | None = Field(default=None, max_length=128, alias="accountIds")

    @field_validator("account_ids")
    @classmethod
    def _validate_unique_account_ids(cls, value: list[str] | None) -> list[str] | None:
        if value is None:
            return None
        normalized = [account_id.strip() for account_id in value if account_id.strip()]
        if len(set(normalized)) != len(normalized):
            raise ValueError("Duplicate account IDs are not allowed")
        return normalized


class AutomationRunResponse(DashboardModel):
    id: str
    job_id: str
    job_name: str | None = None
    model: str | None = None
    reasoning_effort: str | None = None
    trigger: Literal["scheduled", "manual"]
    status: Literal["running", "success", "failed", "partial"]
    scheduled_for: datetime
    started_at: datetime
    finished_at: datetime | None = None
    account_id: str | None = None
    error_code: str | None = None
    error_message: str | None = None
    attempt_count: int = Field(ge=0)
    effective_status: Literal["running", "success", "failed", "partial"] | None = None
    total_accounts: int | None = Field(default=None, ge=0)
    completed_accounts: int | None = Field(default=None, ge=0)
    pending_accounts: int | None = Field(default=None, ge=0)
    cycle_key: str | None = None


class AutomationJobResponse(DashboardModel):
    id: str
    name: str
    enabled: bool
    include_paused_accounts: bool = False
    account_scope_all: bool = Field(default=True, alias="accountScopeAll")
    schedule: AutomationScheduleResponse
    model: str
    reasoning_effort: str | None = None
    prompt: str
    account_ids: list[str]
    next_run_at: datetime | None = None
    last_run: AutomationRunResponse | None = None


class AutomationJobsListResponse(DashboardModel):
    items: list[AutomationJobResponse] = Field(default_factory=list)
    total: int = 0
    has_more: bool = False


class AutomationRunsListResponse(DashboardModel):
    items: list[AutomationRunResponse] = Field(default_factory=list)
    total: int = 0
    has_more: bool = False


class AutomationJobFilterOptionsResponse(DashboardModel):
    account_ids: list[str] = Field(default_factory=list)
    models: list[str] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)
    schedule_types: list[str] = Field(default_factory=list)


class AutomationRunFilterOptionsResponse(DashboardModel):
    account_ids: list[str] = Field(default_factory=list)
    models: list[str] = Field(default_factory=list)
    statuses: list[str] = Field(default_factory=list)
    triggers: list[str] = Field(default_factory=list)


class AutomationRunAccountStateResponse(DashboardModel):
    account_id: str
    status: Literal["pending", "running", "success", "failed", "partial"]
    run_id: str | None = None
    scheduled_for: datetime | None = None
    started_at: datetime | None = None
    finished_at: datetime | None = None
    error_code: str | None = None
    error_message: str | None = None


class AutomationRunDetailsResponse(DashboardModel):
    run: AutomationRunResponse
    accounts: list[AutomationRunAccountStateResponse] = Field(default_factory=list)
    total_accounts: int = Field(ge=0)
    completed_accounts: int = Field(ge=0)
    pending_accounts: int = Field(ge=0)


class AutomationDeleteResponse(DashboardModel):
    status: Literal["deleted"]
